src/py/flwr/cli/log.py

```diff
@@ -26,18 +26,57 @@
 from flwr.cli.config_utils import load_and_validate
 from flwr.common.grpc import GRPC_MAX_MESSAGE_LENGTH, create_channel
 from flwr.common.logger import log as logger
+from flwr.proto.exec_pb2 import StreamLogsRequest  # pylint: disable=E0611
+from flwr.proto.exec_pb2_grpc import ExecStub
 
 CONN_REFRESH_PERIOD = 60  # Connection refresh period for log streaming (seconds)
 
 
-# pylint: disable=unused-argument
-def stream_logs(run_id: int, channel: grpc.Channel, period: int) -> None:
+def stream_logs(run_id: int, channel: grpc.Channel, duration: int) -> None:
     """Stream logs from the beginning of a run with connection refresh."""
+    start_time = time.time()
+    stub = ExecStub(channel)
+    req = StreamLogsRequest(run_id=run_id)
 
+    for res in stub.StreamLogs(req):
+        print(res.log_output)
+        if time.time() - start_time > duration:
+            break
 
-# pylint: disable=unused-argument
-def print_logs(run_id: int, channel: grpc.Channel, timeout: int) -> None:
-    """Print logs from the beginning of a run."""
+
+def print_logs(
+    run_id: int, channel: grpc.Channel, timeout: int, is_test: bool = False
```

Comment:
We can avoid introducing `is_test` here by throwing a keyboard exception in the test.
I'm not so familiar with testing grpc message, but the following test seems to work:

```python
    def test_flwr_log_print_method(self) -> None:
        """Test print_logs."""

        class MockStreamLogsResponse(object):
            @property
            def log_output(self):
                raise KeyboardInterrupt

        mock_obj = Mock()
        mock_obj.StreamLogs.side_effect = (
            [StreamLogsResponse(log_output=f"result_{i}") for i in range(1, 4)]
            + [
                MockStreamLogsResponse(),
            ],
        )

        with patch("flwr.cli.log.ExecStub", return_value=mock_obj):
            with patch("builtins.print") as mock_print:
                print_logs(run_id=123, channel=Mock(), timeout=0)
                mock_print.assert_any_call("result_1")
                mock_print.assert_any_call("result_2")
                mock_print.assert_any_call("result_3")
```

Reply:
Thanks! `Mock.side_effect` is very useful here.